Check the execution receipt received from network properly

liuchengxu opened the pull request:
This PR is mainly for making the locally produced execution receipt persistent for a while so that the executor can compare it with the external ER and propose a fraud proof in case of any invalid state transition in the external ER.

The first 3 commits are basically about some refactorings, and then the core of this PR follows. Making it a draft due to these TODOs, but the other commits are ready to be reviewed.

TODOs:

- [ ] ~~Better approach than using `sleep()` while waiting for the local ER~~, acceptable for now, will have to see if it's usable for the early stage.(Ref https://github.com/subspace/subspace/pull/236#discussion_r795076540).
- [x] Prune the too old ERs.

liuchengxu committed:
Include the list of storage roots in ExecutionReceipt

Refactor `ExecutionReceipt`:

- Rename `state_transition_root` to `trace_root`.
- Include the list of all storage roots in the execution as `trace`.

liuchengxu committed:
Only authorty executors are allowed to broadcast ER

Only the authority executors are allowed to participate the election,
the winner of which will be able to broadcast the exeuction recepit to
the network, and then we could run an authority executor node and
a full executor node locally.

liuchengxu committed:
Reduce Executor generics by merging `runtime_api` into `client`

liuchengxu committed:
Write local ER to disk and check external ER against the local one

liuchengxu committed:
Remove the validated local execution receipt from disk

i1i1 commented on cumulus/client/cirrus-executor/src/aux_schema.rs:
```diff
@@ -0,0 +1,41 @@
+	block_hash: Block::Hash,
+) -> ClientResult<Option<ExecutionReceipt<Block::Hash>>> {
+	load_decode(backend, (EXECUTION_RECEIPT_KEY, block_hash).encode().as_slice())
+}
```
Minor thing, but may be this can be generalized in some manner like this:
```rust
pub(super) struct FromAuxStore<Prefix, T>(PhantomData<Prefix, T>);

impl<Prefix, T> FromAuxStore<Prefix, T>
where
    Prefix: Get<Vec<u8>>, // `impl AsRef<u8>` should also do
    T: Decode,
{
    pub fn write<Backend: AuxStore, Block: BlockT>(
        backend: &Backend,
        block_hash: Block::Hash,
        value: &T,
    ) -> Result<(), sp_blockchain::Error> {
        let key = (Prefix::get().as_ref(), block_hash).encode();
        backend.insert_aux(&[(key.as_slice(), execution_receipt.encode().as_slice())], [])
    }
    // ...
}
```

i1i1 replied:
Getting some key can be  made async with some custom made future!

i1i1 replied:
Something like this:
```rust
// `Get<impl AsRef<u8>>` would suit better
impl<Prefix: Get<Vec<u8>>, T: Decode> FromAuxStore<Prefix, T> {
	pub fn load_async<Backend: AuxStore, Block: BlockT>(
		backend: &'_ Backend,
		block_hash: Block::Hash,
	) -> AuxLoadFuture<'_, Prefix, T, Backend, Block> {
		AuxLoadFuture { backend, block_hash }
	}
}

pub(crate) struct AuxLoadFuture<'a, Prefix, T, Backend, Block: BlockT> {
	backend: &'a Backend,
	block_hash: Block::Hash,
}

impl<'a, Prefix: Get<Vec<u8>>, T: Decode, Backend: AuxStore, Block: BlockT> Future for AuxLoadFuture {
	type Output = ClientResult<T>;
	fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
		match load_decode(self.backend, (Prefix::get().as_ref(), block_hash).encode_as_slice()) {
			Some(result) => Poll::Ready(result),
			None => Poll::Pending,
		}
	}
}
```

liuchengxu replied:
Concerning the mere plain get/set function, I'm reluctant to introduce more abstractions before it's really necessary.

liuchengxu replied:
Making it async is better than `sleep()`, is it ok for now @nazar-pc?

nazar-pc replied:
1. I don't think it needs to be generalized, not at this point for sure.
2. I'm really not used to writing manual futures unless absolutely needed, but above future example isn't a future, it doesn't ever notify executor that it needs to be polled again

i1i1 replied:
> 2. I'm really not used to writing manual futures unless absolutely needed, but above future example isn't a future, it 
doesn't ever notify executor that it needs to be polled again

True, makes sense. Didn't think that you must to wake future. Then we can leave it as it is :)

i1i1 commented on cumulus/client/cirrus-executor/src/lib.rs:
```diff
@@ -388,22 +399,78 @@ where
+			local_receipt
 		} else {
+			// Wait for the local execution receipt until it's ready.
+			let (tx, rx) = crossbeam::channel::bounded::<
+				sp_blockchain::Result<ExecutionReceipt<Block::Hash>>,
+			>(1);
```
Is this the case for a oneshot channel?
https://docs.rs/tokio/1.10.0/tokio/sync/oneshot/index.html

i1i1 replied:
Or is it here in order to block on it in sync context and not to block in async future?

liuchengxu replied:
Yeah, it's the oneshot case. The problem here is how to properly get a value asynchronously under a sync context. The background is that when an executor received an execution receipt from the network, the local execution receipt might be not ready and you have to wait because you can't check the external execution receipt until you have generated it on your own locally.

nazar-pc replied:
I think it is acceptable for now.

I traced call stack and it is called from manually written future in Substrate's source code. So while we could have refactored everything all the way up there to make it async, it is unlikely to be accepted and from one of the PRs Liu-Cheng sent upstream it looks like there is a chance this gossiping mechanism will go away eventually.

We'll likely switch to our own networking layer for these kinds of things once it is ready.

Just make sure this doesn't cause disasters as it does seem to block networking completely.

i1i1 commented on cumulus/client/cirrus-executor/src/lib.rs:
```diff
@@ -388,22 +399,78 @@ where
+								let _ = tx.send(Ok(local_receipt));
+								break
+							},
+							Ok(None) => {
+								tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+							},
```
Going into the aux storage can be made async with some premade future. Left a comment with example above.

nazar-pc commented on cumulus/client/cirrus-executor/src/processor.rs:
```diff
@@ -195,28 +195,27 @@ where
 		roots.push(state_root.encode());
 
+		let trace_root =
+			BlakeTwo256::ordered_trie_root(roots.clone(), sp_core::storage::StateVersion::V1);
```
This is not a regular Merkle Tree though, despite `trace_root` claiming the opposite in `ExecutionReceipt` data structure...

liuchengxu replied:
Hmm, what's a regular Merkle Tree? And the suggestion here is?

nazar-pc replied:
Merkle Tree is Merkle Tree. Merkle Particia Trie (that you used there) is a different (but related) data structure.

liuchengxu replied:
Sure, I see the point. Usually, I mix them up in practice unless it's necessary to distinguish because you could use both of them for some common purposes. So, you want it to be a plain Merkle Tree not a MPT here?

nazar-pc replied:
Yes, that was the design we went for unless I missed something. Do we have other places where we use MPT for similar purpose?

liuchengxu replied:
This is the only one for now. Does using MPT have some downsides? Perhaps we could use it since substrate already provides this convenient API if there is no harm doing so.

nazar-pc replied:
MPT is a different data structure used for storing data in it. MT is not. We don't need some ordered version of that with layout V1, we need a simple MT instead. I understand that we have this convenient API to get some root, but it is not what we need.

liuchengxu replied:
Please take a look at https://github.com/subspace/subspace/pull/236/commits/e0631d3df11b847135019ae4bf64a70e548903fe

nazar-pc commented on cumulus/client/cirrus-executor/src/lib.rs:
```diff
@@ -90,36 +89,35 @@ impl<Block: BlockT, BS, RA, Client, TransactionPool, Backend, CIDP> Clone
 			execution_receipt_sender: self.execution_receipt_sender.clone(),
 			backend: self.backend.clone(),
 			create_inherent_data_providers: self.create_inherent_data_providers.clone(),
+			is_authority: self.is_authority.clone(),
```
```suggestion
			is_authority: self.is_authority,
```

nazar-pc commented on cumulus/client/cirrus-executor/src/lib.rs:
```diff
@@ -388,22 +399,78 @@ where
+			rx.recv()??
+		};
+
+		// TODO: What happens for this obvious error?
```
Same as in any other case, you compare trace item one by one and generate fraud proof for item that is wrong or missing

liuchengxu replied:
There is a possible case that an external ER that contains all the valid roots and some extra ones? In that case comparing on the iterator won't detect it, but I believe the block will still fail in the verification somewhere else.

nazar-pc replied:
Everything is possible, adversary can create whatever ER they want :slightly_smiling_face:

nazar-pc commented on cumulus/client/cirrus-executor/src/lib.rs:
```diff
@@ -388,22 +399,78 @@ where
+				.trace
+				.iter()
+				.enumerate()
+				.zip(execution_receipt.trace.iter().enumerate())
```
Does this need to be enumerated though? You ignore the index above anyway.

liuchengxu replied:
Yeah, the index of external ER is ignored, but the local index is returned. Although the storage proof has not been implemented, I guess we'll need the index for generating the proof?

liuchengxu replied:
So far, I see the index of local receipt will be necessary for generating the execution proof in either way mentioned in https://github.com/paritytech/substrate/discussions/10783.

nazar-pc commented on cumulus/client/cirrus-executor/src/lib.rs:
```diff
@@ -388,22 +399,78 @@ where
+			None
+		};
+
+		if let Some(_local_trace_idx) = find_invalid_transition() {
```
I'd use `.find()` on iterator instead of wrapping into a function with `return` to make it more idiomatic

liuchengxu replied:
Updated.

nazar-pc commented on cumulus/client/cirrus-executor/src/processor.rs:
```diff
@@ -216,6 +216,12 @@ where
+			&*self.client,
+			header_hash,
+			&execution_receipt,
+		)?;
```
TODO is needed that this may be never cleaned up (as we discussed in DMs)

liuchengxu replied:
https://github.com/subspace/subspace/pull/236/commits/2587e806d8ce5506352d49f39a229c5196d302c6 Please see if it could serve as a starting point.

nazar-pc reviewed (commented):
Overall, I think we need something like orphan ERs in this case to handle it gracefully, but it is definitely tricky to do in a way that is DoS resistant

liuchengxu committed:
Apply the suggestions from code review

liuchengxu committed:
Use Merkle Tree instead of Merkle Patricia Tree for execution trace

Fixes https://github.com/subspace/subspace/pull/236#discussion_r795073177

liuchengxu committed:
Prune the too old execution receipts while writing a new one

We have to use the block number as key because the time is measured in
block number only.

nazar-pc commented on cumulus/client/cirrus-executor/Cargo.toml:
```diff
@@ -30,7 +30,9 @@ futures = { version = "0.3.19", features = ["compat"] }
 rand_chacha = "0.3.1"
+merkletree = "0.21.0"
 parking_lot = "0.11.2"
+sha2 = "0.10.0"
```
Check how this dependency is added to other crates to take advantage of platform-specific hardware accelerated instructions

nazar-pc commented on cumulus/client/cirrus-executor/src/processor.rs:
```diff
@@ -194,8 +194,13 @@ where
+				.iter()
+				.map(|r| r.as_slice().try_into().expect("Storage root type is [u8; 32]; qed")),
+		)
+		.expect("Failed to construct merkle tree for execution trace")
```
This should ideally have an explanation why error will never happen

nazar-pc commented on cumulus/client/cirrus-executor/src/processor.rs:
```diff
@@ -194,8 +194,13 @@ where
+		let trace_root = crate::merkle_tree::MerkleTree::new(
+			roots
+				.iter()
+				.map(|r| r.as_slice().try_into().expect("Storage root type is [u8; 32]; qed")),
```
This doesn't give me a lot of confidence. The whole flow throughout is just `Vec<u8>` and only at the last step transformed into `[u8; 32]`. Can we make intermediate roots `[u8; 32]` in the first place instead? That way if there is a mismatch, it will be uncovered much sooner.

nazar-pc commented on cumulus/client/cirrus-executor/src/aux_schema.rs:
```diff
@@ -22,29 +28,140 @@ fn load_decode<Backend: AuxStore, T: Decode>(
+		backend.insert_aux(
+			&[
+				(
+					(EXECUTION_RECEIPT_KEY, block_number).encode().as_slice(),
```
I thought I already commented on this somewhere, but ERs should be stored using block hash, not block number. Otherwise we'll get issues when handling forks of the primary chain.

liuchengxu replied:
I have mentioned more in the commit message, the only way to measure time is the block number in blockchain, you can't use a hash to decide if it's too old.

nazar-pc replied:
Definitely, but that doesn't mean we should use block number anyway and have issues with forks. This is exactly why the issue with leaked storage wasn't fixes quickly in BABE, it is tricky. BTW there is a draft PR (that I didn't read in much detail) that aims to deal with this: https://github.com/paritytech/substrate/pull/10748

liuchengxu replied:
What's more, I'm currently not sure what's the way to handle forks in cirrus, do we just wait for a confirmation time like bitcoin or it's something else?

nazar-pc replied:
No, we don't need to handle forks in any way other than following primary chain. If primary chain forks - secondary chain forks as well. If outdated branches are pruned - corresponding branches are pruned as well. This is one of the reasons why I said that this is unlike the way parachains work, execution deterministically follows primary chain.

liuchengxu replied:
In order to store the receipt by hash, we need another storage to track the associated block number, otherwise, we can't decide if the receipt is too old and should be pruned. In addition, we should not delete the local receipt even it's validated correctly, otherwise if somehow the executor received the same valid receipt again, the handler will hang forever. That means the deletion of execution receipts will rely on the pruning depth only for now, but at least it won't increase without a limit.

nazar-pc commented on cumulus/client/cirrus-executor/src/aux_schema.rs:
```diff
@@ -0,0 +1,167 @@
+const EXECUTION_RECEIPT_KEY: &[u8] = b"execution_receipt";
+const EXECUTION_RECEIPT_START: &[u8] = b"execution_receipt_start";
+/// Prune the execution receipts when they reach this number.
+const PRUNING_BOUND: u64 = 1000;
```
This would be a better name:
```suggestion
const PRUNING_DEPTH: u64 = 1000;
```

liuchengxu committed:
Store Execution Receipt Keyed by block hash

Store the execution receipt by block hash, use another storage for
tracking the associated block number to have knowledge of if it's too
old and should be pruned.

This commit also removes the deletion of local ER even it's correctly
validated, now the obsolete ERs will be removed soly according to the
prune depth.

liuchengxu committed:
Replicate the state root when constructing the trace merkle tree from the root list

liuchengxu committed:
Merge the `BS` generic parameter into `Client` in Executor struct

liuchengxu committed:
Use the new best_hash instead of parent_hash when calling `intermediate_roots()`

Using `parent_hash` is wrong because we need to call the api on top of
the newly created best block to get the intermediate roots for this
just executed block.

nazar-pc commented on cumulus/client/cirrus-executor/src/aux_schema.rs:
```diff
@@ -47,18 +59,28 @@ pub(super) fn write_execution_receipt<Backend: AuxStore, Block: BlockT>(
-			keys_to_delete.push((EXECUTION_RECEIPT_KEY, to_delete_start).encode());
-			to_delete_start = to_delete_start.saturating_add(One::one());
+		let mut to_delete_block_number = first_saved_receipt;
+		while to_delete_block_number < new_first_saved_receipt {
```
`to_delete_block_number` is basically `from` and `new_first_saved_receipt` is basically `to`.

Instead of transforming those into `Block::Header::Number`, I'd convert them into `subspace_core_primitives::BlockNumber` and iterated over the range of numbers instead, that would be easier to read rather than manually incrementing stuff.

liuchengxu replied:
Converting to `subspace_core_primitives::BlockNumber` might be helpful to simplify the loop, but you have to convert it back to `Number` as the encode for `u32` and `Block::Header::Number` is different(discovered by the existing test, perhaps  the compact of codec is related, I didn't look into this) unless you want to use `u32` as the part of the final key to insert into the store.

nazar-pc replied:
I think it is fine to use `BlockNumber` as a final key. Also it might be different in tests (I see that often tests use `u64` for block number), but should be the same in production since we use `u32` everywhere as a block number otherwise.

nazar-pc commented on cumulus/client/cirrus-executor/src/aux_schema.rs:
```diff
@@ -47,18 +59,28 @@ pub(super) fn write_execution_receipt<Backend: AuxStore, Block: BlockT>(
+			to_delete_block_number = to_delete_block_number.saturating_add(One::one());
 		}
 
 		backend.insert_aux(
```
This insert has the first argument repeated in both branches, maybe makes sense to move it out of the `if`, especially since code is a bit verbose.

nazar-pc commented on cumulus/client/cirrus-executor/src/aux_schema.rs:
```diff
@@ -47,18 +59,28 @@ pub(super) fn write_execution_receipt<Backend: AuxStore, Block: BlockT>(
 		new_first_saved_receipt = block_number.saturating_sub((PRUNING_DEPTH - 1).saturated_into());
```
In these cases I would honestly prefer `.expect()` since we don't expect these numbers to ever be beyond `u32::MAX`.

liuchengxu replied:
Feel free to refactor it directly as I don't see another style that is satisfying to me using `expect()` :P.

nazar-pc commented on cumulus/client/cirrus-executor/src/processor.rs:
```diff
@@ -192,36 +194,46 @@ where
+		let state_root = state_root
+			.encode()
+			.try_into()
+			.expect("State root uses the same Block hash type which can fit into [u8; 32]; qed");
```
https://www.ietf.org/rfc/rfc2119.txt
```suggestion
			.expect("State root uses the same Block hash type which must fit into [u8; 32]; qed");
```

liuchengxu replied:
I have no problem following that practice, just want to say:

1. It's not an RFC in any way.
2. It's not universally applied in our codebase with a quick search.
    - https://github.com/subspace/subspace/blob/5363a50c0491e75278fd357a1e63efe1ec83b9f3/crates/subspace-archiving/src/reconstructor.rs#L109
    - https://github.com/subspace/subspace/blob/5363a50c0491e75278fd357a1e63efe1ec83b9f3/crates/subspace-archiving/src/archiver.rs#L244
    - https://github.com/subspace/subspace/blob/5363a50c0491e75278fd357a1e63efe1ec83b9f3/crates/subspace-archiving/src/archiver.rs#L637
    - https://github.com/subspace/subspace/blob/5363a50c0491e75278fd357a1e63efe1ec83b9f3/crates/sc-consensus-subspace/src/archiver.rs#L40
    - https://github.com/subspace/subspace/blob/5363a50c0491e75278fd357a1e63efe1ec83b9f3/crates/pallet-subspace/src/lib.rs#L526

nazar-pc replied:
Then we need to fix those as well. It is generally a good practice to use "should" for things that are usually the way they are described and "must" when they absolutely have to be.

liuchengxu replied:
I agree. Furthermore, I think it's beneficial to also introduce a code-style doc starting with fixing them since the team is growing and we're having more and more unwritten rules.

nazar-pc replied:
I would be thankful if you start that initiative

nazar-pc commented on cumulus/client/cirrus-executor/src/merkle_tree.rs:
```diff
@@ -0,0 +1,67 @@
+
+	let roots_len = roots.len();
+	// roots contains at least [storage_root_after_initializing_block, state_root].
+	assert!(roots_len >= 2, "Execution trace should at least contain 2 storage roots");
```
https://www.ietf.org/rfc/rfc2119.txt
```suggestion
	assert!(roots_len >= 2, "Execution trace must at least contain 2 storage roots");
```

nazar-pc commented on cumulus/client/cirrus-executor/src/merkle_tree.rs:
```diff
@@ -41,3 +41,26 @@ pub(super) type MerkleTree = merkletree::merkle::MerkleTree<
+		if let Some(state_root) = roots.last().copied() {
+			roots.resize(ideal_len, state_root);
+		}
+	}
```
Hm... doesn't library already handle this though?

liuchengxu replied:
No, user has to take care of that on their own https://docs.rs/merkletree/0.21.0/src/merkletree/merkle.rs.html#1705

nazar-pc reviewed (commented):
Some comments, but nothing critical

liuchengxu committed:
Apply the suggstions from code review
